components/ledger/pkg/machine/script/compiler/compiler.go

```diff
@@ -62,35 +68,56 @@ func (p *parseVisitor) isWorld(addr core.Address) bool {
 	return false
 }
 
-func (p *parseVisitor) VisitVariable(c parser.IVariableContext, push bool) (core.Type, *core.Address, *CompileError) {
+func (p *parseVisitor) VisitVariable(c parser.IVariableContext, push bool) (*ExprInfo, *CompileError) {
 	name := c.GetText()[1:] // strip '$' prefix
 	if idx, ok := p.varIdx[name]; ok {
 		res := p.resources[idx]
 		if push {
 			p.PushAddress(idx)
 		}
-		return res.GetType(), &idx, nil
+		var assetSource *core.Address
+		if res.GetType() == core.TypeMonetary ||
+			res.GetType() == core.TypeAsset {
+			assetSource = &idx
+		}
+		return &ExprInfo{
+			typ:         res.GetType(),
+			addr:        &idx,
+			assetSource: assetSource,
```

Comment:
I need to store the address of the asset only when the variable is of type `asset` or `monetary`

Reply:
Ah just saw this, I mentioned [in this comment](https://github.com/formancehq/stack/pull/215#discussion_r1169299821) why I added it, it's because when doing `$mon_a + $mon_b`, there is no resource address (the expression's result is not a resource), however we still need the corresponding asset for the `send` mechanism, so as a hack we get the `assetSource` of one of the operands.
Maybe it's not too big of a deal, but I'm afraid that if we use the `addr` field for this, in the future we might accidentally end up pushing the `addr` thinking that we just pushed the expression's result, and it will silently push the wrong value without complaining.